Mage.Sets/src/mage/cards/t/TheStoneBrain.java

```diff
@@ -0,0 +1,128 @@
+                Zone.BATTLEFIELD, new ChooseACardNameEffect(ChooseACardNameEffect.TypeOfName.ALL), new GenericManaCost(2)
+        );
+        ability.addEffect(new TheStoneBrainEffect());
+        ability.addTarget(new TargetOpponent());
+        ability.addCost(new TapSourceCost());
+        ability.addCost(new ExileSourceCost());
+        this.addAbility(ability);
+    }
+
+    private TheStoneBrain(final TheStoneBrain card) {
+        super(card);
+    }
+
+    @Override
+    public TheStoneBrain copy() {
+        return new TheStoneBrain(this);
+    }
+}
+
+class TheStoneBrainEffect extends OneShotEffect {
+
+    public TheStoneBrainEffect() {
+        super(Outcome.Benefit);
+        this.staticText = "Search target opponent's graveyard, hand, and library for up to four cards with that name and exile them. That player shuffles, then draws a card for each card exiled from their hand this way";
+    }
+
+    public TheStoneBrainEffect(final TheStoneBrainEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public TheStoneBrainEffect copy() {
+        return new TheStoneBrainEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        String cardName = (String) game.getState().getValue(source.getSourceId().toString() + ChooseACardNameEffect.INFO_KEY);
+        Player controller = game.getPlayer(source.getControllerId());
+        if (cardName != null && controller != null) {
+            int numberOfCardsStillToRemove = 4;
+            int numberOfCardsExiledFromHand = 0;
+            Player targetPlayer = game.getPlayer(getTargetPointer().getFirst(game, source));
+            if (targetPlayer != null) {
+                FilterCard filter = new FilterCard("card named " + cardName);
+                filter.add(new NamePredicate(cardName));
+
+                // cards in Graveyard
+                int cardsCount = (cardName.isEmpty() ? 0 : targetPlayer.getGraveyard().count(filter, game));
+                if (cardsCount > 0) {
+                    filter.setMessage("card named " + cardName + " in the graveyard of " + targetPlayer.getName());
+                    TargetCard target = new TargetCard(0, Math.min(cardsCount, numberOfCardsStillToRemove), Zone.GRAVEYARD, filter);
+                    if (controller.choose(Outcome.Exile, targetPlayer.getGraveyard(), target, game)) {
+                        numberOfCardsStillToRemove -= target.getTargets().size();
+                        controller.moveCards(new CardsImpl(target.getTargets()), Zone.EXILED, source, game);
+                    }
+                }
+
+                // cards in Hand
+                cardsCount = (cardName.isEmpty() ? 0 : targetPlayer.getHand().count(filter, game));
+                filter.setMessage("card named " + cardName + " in the hand of " + targetPlayer.getName());
+                TargetCard target = new TargetCard(0, Math.min(cardsCount, numberOfCardsStillToRemove), Zone.HAND, filter);
```

Comment:
Yes. Because all of the looping happens within the TargetCard function I don't think there's anything I can do to change how it works externally, so would have to make custom effects / targets for all cards bugged by the 0 magic number. Given that it frequently causes bugs and doesn't seem(?) to have any use case (see my comment [here](https://github.com/magefree/mage/pull/9727#discussion_r1026839274)) I feel like it should be changed in `TargetCard` as opposed to making attempts at circumventing the issue every time it comes up

Reply:
> Yes. Because all of the looping happens within the TargetCard function I don't think there's anything I can do to change how it works externally, so would have to make custom effects / targets for all cards bugged by the 0 magic number. Given that it frequently causes bugs and doesn't seem(?) to have any use case (see my comment [here](https://github.com/magefree/mage/pull/9727#discussion_r1026839274)) I feel like it should be changed in `TargetCard` as opposed to making attempts at circumventing the issue every time it comes up

Yes, I agree with you.

Reply:
@awjackson, @PurpleCrowbar, @Alex-Vasile, is The Stone Brain currently performing well enough to be added to Xmage Beta? Silence since November...